### PR TITLE
Split files & nits

### PR DESCRIPTION
**Pull request by fstamour**
split `cluck.lisp` into multiple files:

- analysis.lisp
- checks.lisp
- egraph.lisp
- plot.lips
- rewrites.lisp
- saturation.lisp

### CODE REVIEW

**Comment by fstamour** [comment]
- [x] put the notices at the beginning of the files
  - `;;;;` instead of `;;`
  - `Author` instead of `Copyright`
- [x] rebase

File: src/analysis.lisp [hunk 1]
```diff
@@ -0,0 +1,88 @@
+(in-package :cluck)
```
Comment (stylewarning) [inline src/analysis.lisp]: use #:

File: src/analysis.lisp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+(defclass e-analyzer-cost (e-analyzer)
+  ()
+  (:documentation "An e-class analyzer which can be used as the cost function when determining the best concrete representation of an e-class. Should implement E-CLASS-ANALYZER-COMPARE. A canned implementation of E-ANALYZER-MERGE exists for instances, which simply finds the cheapest e-node according to the analysis value; this implementation must not be overridden."))
+
+(defgeneric e-analyzer-e-node-value (analyzer e-node-car children)
+  (:documentation "Return the analysis value for an e-node. CHILDREN is a list of analysis values of the children. Some elements of CHILDREN may have circular references back to the e-class containing this e-node. In that case, E-ANALYZER-PARENT-VALUE is called to determine a reasonable default value. Note that in some situations, a circular reference will not use the parent value, depending on order of traversal, and instead will just be the value not including this new node."))
+(defgeneric e-analyzer-compare (analyzer v1 v2)
```
Comment (stylewarning) [inline src/analysis.lisp:15]: newline

File: src/analysis.lisp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+(defgeneric e-analyzer-parent-value (analyzer)
+  (:documentation "See E-ANALYZER-E-NODE-VALUE. Generally speaking, should return some logically maximum/high-cost value for cost analyzers, since a circular reference has infinite size.")
+  (:method (a)
+    'parent))
+
+(defclass tree-size-analyzer (e-analyzer-cost)
+  ()
+  (:documentation "An e-class analyzer that simply counts the smallest representation of each e-class."))
+(defvar +tree-size-analyzer+ (make-instance 'tree-size-analyzer))
```
Comment (stylewarning) [inline src/analysis.lisp:32]: newlines

File: src/checks.lisp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                    :do (error "Congruence invariant violated: Multiple e-nodes congruent to ~a." canonical-en))))
+
+(declftype (e-graph) boolean check-e-graph)
+(defun check-e-graph (eg)
+  "Throw a error if the given e-graph violates its key invariants."
+  (check-e-graph-hashcons-invariant eg)
+  (check-e-graph-congruence-invariant eg)
+  t)
+
+;; TODO: add a function to print all representations of an e-class. Will be a bit annoying to deal with circular graphs.
```
Comment (stylewarning) [inline src/checks.lisp]: turn this into an issue

File: src/egraph.lisp [hunk 5]
```diff
@@ -0,0 +1,277 @@
+;; Copyright (c) Mark Polyakov, released under MIT License
+
+(in-package :cluck)
+
+;; An enode is represented as a list whose CAR is a symbol denoting a function and whose CDR is a list of e-class IDs
```
Comment (stylewarning) [inline src/egraph.lisp]: ;;; and M-q

**Review by stylewarning — CHANGES_REQUESTED** [review]
all good, a few nits + your TODOs

File: src/analysis.lisp (hunk 1 above)
Comment (fstamour) [reply src/analysis.lisp]: done

File: src/checks.lisp (hunk 4 above)
Comment (fstamour) [reply src/checks.lisp]: done

File: src/egraph.lisp (hunk 5 above)
Comment (fstamour) [reply src/egraph.lisp]: done

File: src/analysis.lisp (hunk 3 above)
Comment (fstamour) [reply src/analysis.lisp:32]: done

(those where like that before I moved them :laughing:)

File: src/analysis.lisp (hunk 2 above)
Comment (fstamour) [reply src/analysis.lisp:15]: done

File: src/analysis.lisp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+(defclass e-analyzer-cost (e-analyzer)
+  ()
+  (:documentation "An e-class analyzer which can be used as the cost function when determining the best concrete representation of an e-class. Should implement E-CLASS-ANALYZER-COMPARE. A canned implementation of E-ANALYZER-MERGE exists for instances, which simply finds the cheapest e-node according to the analysis value; this implementation must not be overridden."))
+
+(defgeneric e-analyzer-e-node-value (analyzer e-node-car children)
+  (:documentation "Return the analysis value for an e-node. CHILDREN is a list of analysis values of the children. Some elements of CHILDREN may have circular references back to the e-class containing this e-node. In that case, E-ANALYZER-PARENT-VALUE is called to determine a reasonable default value. Note that in some situations, a circular reference will not use the parent value, depending on order of traversal, and instead will just be the value not including this new node."))
+(defgeneric e-analyzer-compare (analyzer v1 v2)
```
Comment (stylewarning) [inline src/analysis.lisp:15]: ideally we put new lines between all definition forms. we can punt that until later.

File: src/package.lisp [hunk 7]
```diff
@@ -1,67 +1,69 @@
-(uiop:define-package :CLuck
-  (:use :cl
-        :alexandria)
+;;;; Author: Mark Polyakov, released under MIT License
+
+(uiop:define-package #:cluck
```
Comment (stylewarning) [inline src/package.lisp:3]: feelsgoodman.jpeg

File: src/plot.lisp [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+(defun prin1-e-node-to-string (to-printable en)
+  (prin1-to-string (cons (funcall to-printable (car en))
+                         (cdr en))))
+
+(declftype (e-graph &key (:stream (or stream (eql t))) (:e-node-car-to-printable function)) t e-graph-print)
+(defun e-graph-print (eg &key (stream t) (e-node-car-to-printable #'identity))
+  (let ((num-dirty-ens (hash-table-count (e-graph-dirty-e-nodes eg)))
+        (ids (e-graph-e-class-id-list eg)))
+    (when (not (zerop num-dirty-ens))
+      (warn "WARNING: The e-graph is dirty! (~a many dirty nodes).~%~%" num-dirty-ens))
```
Comment (stylewarning) [inline src/plot.lisp:15]: later: we shouldn't write WARNING:

**Review by stylewarning — APPROVED** [review]

**Pull request merged**